### PR TITLE
Support --dry-run -o yaml

### LINKED ISSUE #95: Add ability to edit the kubernetes manifests
It would be nice if the user was able to inspect/modify the manifests deployed with the `kubeapps up` command. A possible solution is to introduce a `kubeapps dump` command to dump the complete manifest yaml which could be then edited and deployed using `kubectl`.

This would be helpful in getting kubeapps installed on clusters where the default kubeapps manifest doesn't "just work".

Sample usage:

```bash
$ kubeapps dump | kubectl create -f -
```

### PR DESCRIPTION
**Pull request by ngtuna**
**Description** This PR fixes #95 which is adding support for printing out the embedded manifest in both yaml and json format. The command is `kubeapps up --dry-run -o yaml`

- [x] implement
- [x] documentation
- [x] unit test

### CODE REVIEW

File: cmd/up.go [hunk 1]
```diff
@@ -170,7 +182,52 @@ List of components that kubeapps up installs:
...
+		for _, obj := range objs {
+			j, err := json.MarshalIndent(obj, "", "    ")
+			if err != nil {
+				return fmt.Errorf("can't dump kubeapps manifest: %v", err)
+			}
+			bObjs = append(bObjs, j)
+		}
+
+		b := bytes.Join(bObjs, []byte(fmt.Sprintf(",\n")))
+		// insert '[' to the front
```
Comment (prydonius) [inline cmd/up.go]: this isn't the correct format kubectl accepts for multiple json resources (though it's correct json). It should just be json objects separated by a single new line e.g. https://github.com/kubernetes/kubernetes/blob/master/hack/testdata/multi-resource-json.json.

I was able to install the resources with `kubectl` after removing the appending/prepending and just having `b := bytes.Join(bObjs, []byte(fmt.Sprintf("\n")))`

File: cmd/up.go [hunk 2]
```diff
@@ -170,7 +182,52 @@ List of components that kubeapps up installs:
 
 func init() {
 	RootCmd.AddCommand(upCmd)
-	upCmd.Flags().Bool("dry-run", false, "Provides output to be submitted to the server.")
+	upCmd.Flags().Bool("dry-run", false, "Show manifest to be submitted to the k8s cluster without deploying.")
+	upCmd.Flags().StringP("out", "o", "yaml", "Provides manifest format: yaml | json")
```
Comment (prydonius) [inline cmd/up.go]: I would add that this is only used for `--dry-run`

File: cmd/up_test.go [hunk 3]
```diff
@@ -194,3 +195,28 @@ func TestBuildSecret(t *testing.T) {
...
+	objs = append(objs, obj)
+	var buf bytes.Buffer
+
+	err := dump(&buf, "yaml", objs)
+	if err != nil {
+		t.Error(err)
+	}
+	output := buf.String()
+	t.Log("output is", output)
+	if !strings.Contains(output, "foo") || !strings.Contains(output, "baz") {
```
Comment (prydonius) [inline cmd/up_test.go]: I would create test yaml and json output files for this test data and test it matches exactly

File: cmd/up.go (hunk 1 above)
Comment (ngtuna) [reply cmd/up.go]: 👍 Didn't notice that kubectl-style format

File: cmd/up.go (hunk 2 above)
Comment (ngtuna) [reply cmd/up.go]: In fact `-o` defaults to `yaml` so only `--dry-run` will also work. But yes, `-o` is used only for `--dry-run`

**Comment by ngtuna** [comment]
@prydonius 

> not sure if kubecfg has something

kubecfg has its own order strategy. Look at https://github.com/ksonnet/kubecfg/blob/v0.5.0/utils/sort.go#L45

I just added a commit to add the order for generated manifest, and also fix json kubectl-style format. It's good to review now

File: cmd/up.go (hunk 2 above)
Comment (prydonius) [reply cmd/up.go]: what I mean is, this has no effect if I use `-o` without `--dry-run`, which might be confusing? So for that reason just changing the description to something like: "Dry-run manifest output format: yaml | kson" would clear that up IMO

File: cmd/up.go [hunk 4]
```diff
@@ -170,7 +183,45 @@ List of components that kubeapps up installs:
...
+}
+
+func dump(w io.Writer, out string, objs []*unstructured.Unstructured) error {
+	bObjs := [][]byte{}
+	sort.Sort(utils.DependencyOrder(objs))
```
Comment (prydonius) [inline cmd/up.go]: sweet!

**Comment by ngtuna** [comment]
Thanks @prydonius . Fixed that. It's good to be merged now.

**Comment by sameersbn** [comment]
@ngtuna tried the following on a GKE cluster with RBAC enabled.

```console
# kubeapps up --dry-run | kubectl create -f -
namespace "kubeless" created
namespace "kubeapps" created
configmap "kubeapps-dashboard-api-9712a3a" created
clusterrolebinding "tiller-cluster-admin" created
serviceaccount "tiller" created
configmap "kubeapps-dashboard-ui-config-024dc17" created
deployment "kubeapps-dashboard-ui" created
service "kubeapps-dashboard-ui" created
configmap "kubeapps-dashboard-ui-vhost-425de41" created
ingress "kubeapps" created
service "broker" created
customresourcedefinition "sealedsecrets.bitnami.com" created
service "kafka" created
service "kubeapps-dashboard-api" created
deployment "kubeless-controller" created
clusterrolebinding "kubeless-controller-deployer" created
service "zookeeper" created
serviceaccount "controller-acct" created
customresourcedefinition "functions.k8s.io" created
service "zoo" created
deployment "kubeless-ui" created
clusterrolebinding "kubeless-ui-editor" created
serviceaccount "kubeless-ui" created
service "kubeless-ui" created
deployment "mongodb" created
persistentvolumeclaim "mongodb-data" created
service "mongodb" created
configmap "tcp-services" created
deployment "default-http-backend" created
service "default-http-backend" created
clusterrolebinding "nginx-ingress-controller" created
rolebinding "nginx-ingress-controller" created
service "nginx-ingress" created
serviceaccount "nginx-ingress-controller" created
configmap "udp-services" created
configmap "nginx-ingress" created
deployment "nginx-ingress-controller" created
deployment "kubeapps-dashboard-api" created
serviceaccount "sealed-secrets-controller" created
deployment "sealed-secrets-controller" created
service "sealed-secrets-controller" created
rolebinding "sealed-secrets-controller" created
clusterrolebinding "sealed-secrets-controller" created
secret "mongodb" created
statefulset "zoo" created
statefulset "kafka" created
Error from server (Forbidden): error when creating "STDIN": clusterroles.rbac.authorization.k8s.io "secrets-unsealer" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:["sealedsecrets"], APIGroups:["bitnami.com"], Verbs:["get"]} PolicyRule{Resources:["sealedsecrets"], APIGroups:["bitnami.com"], Verbs:["list"]} PolicyRule{Resources:["sealedsecrets"], APIGroups:["bitnami.com"], Verbs:["watch"]} PolicyRule{Resources:["secrets"], APIGroups:[""], Verbs:["create"]} PolicyRule{Resources:["secrets"], APIGroups:[""], Verbs:["update"]} PolicyRule{Resources:["secrets"], APIGroups:[""], Verbs:["delete"]}] user=&{sameer@bitnami.com  [system:authenticated] map[]} ownerrules=[PolicyRule{Resources:["selfsubjectaccessreviews"], APIGroups:["authorization.k8s.io"], Verbs:["create"]} PolicyRule{NonResourceURLs:["/api" "/api/*" "/apis" "/apis/*" "/healthz" "/swaggerapi" "/swaggerapi/*" "/version"], Verbs:["get"]}] ruleResolutionErrors=[]
Error from server (Forbidden): error when creating "STDIN": clusterroles.rbac.authorization.k8s.io "kubeless-controller-deployer" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["create"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["delete"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["update"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["patch"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["create"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["delete"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["update"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["patch"]} PolicyRule{Resources:["deployments"], APIGroups:["apps"], Verbs:["create"]} PolicyRule{Resources:["deployments"], APIGroups:["apps"], Verbs:["get"]} PolicyRule{Resources:["deployments"], APIGroups:["apps"], Verbs:["delete"]} PolicyRule{Resources:["deployments"], APIGroups:["apps"], Verbs:["list"]} PolicyRule{Resources:["deployments"], APIGroups:["apps"], Verbs:["update"]} PolicyRule{Resources:["deployments"], APIGroups:["apps"], Verbs:["patch"]} PolicyRule{Resources:["deployments"], APIGroups:["extensions"], Verbs:["create"]} PolicyRule{Resources:["deployments"], APIGroups:["extensions"], Verbs:["get"]} PolicyRule{Resources:["deployments"], APIGroups:["extensions"], Verbs:["delete"]} PolicyRule{Resources:["deployments"], APIGroups:["extensions"], Verbs:["list"]} PolicyRule{Resources:["deployments"], APIGroups:["extensions"], Verbs:["update"]} PolicyRule{Resources:["deployments"], APIGroups:["extensions"], Verbs:["patch"]} PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["delete"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["get"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["list"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["watch"]}] user=&{sameer@bitnami.com  [system:authenticated] map[]} ownerrules=[PolicyRule{Resources:["selfsubjectaccessreviews"], APIGroups:["authorization.k8s.io"], Verbs:["create"]} PolicyRule{NonResourceURLs:["/api" "/api/*" "/apis" "/apis/*" "/healthz" "/swaggerapi" "/swaggerapi/*" "/version"], Verbs:["get"]}] ruleResolutionErrors=[]
Error from server (Forbidden): error when creating "STDIN": clusterroles.rbac.authorization.k8s.io "kubeless-editor" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["pods/log"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["pods/log"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["get"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["list"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["watch"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["create"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["patch"]} PolicyRule{Resources:["functions"], APIGroups:["k8s.io"], Verbs:["delete"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["proxy"]} PolicyRule{Resources:["services/proxy"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["services/proxy"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["services/proxy"], APIGroups:[""], Verbs:["proxy"]}] user=&{sameer@bitnami.com  [system:authenticated] map[]} ownerrules=[PolicyRule{Resources:["selfsubjectaccessreviews"], APIGroups:["authorization.k8s.io"], Verbs:["create"]} PolicyRule{NonResourceURLs:["/api" "/api/*" "/apis" "/apis/*" "/healthz" "/swaggerapi" "/swaggerapi/*" "/version"], Verbs:["get"]}] ruleResolutionErrors=[]
Error from server (Forbidden): error when creating "STDIN": roles.rbac.authorization.k8s.io "nginx-ingress-controller" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["secrets"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["namespaces"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["configmaps"], ResourceNames:["ingress-controller-leader-nginx"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["configmaps"], ResourceNames:["ingress-controller-leader-nginx"], APIGroups:[""], Verbs:["update"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["create"]} PolicyRule{Resources:["endpoints"], APIGroups:[""], Verbs:["get"]}] user=&{sameer@bitnami.com  [system:authenticated] map[]} ownerrules=[PolicyRule{Resources:["selfsubjectaccessreviews"], APIGroups:["authorization.k8s.io"], Verbs:["create"]} PolicyRule{NonResourceURLs:["/api" "/api/*" "/apis" "/apis/*" "/healthz" "/swaggerapi" "/swaggerapi/*" "/version"], Verbs:["get"]}] ruleResolutionErrors=[]
Error from server (Forbidden): error when creating "STDIN": clusterroles.rbac.authorization.k8s.io "nginx-ingress-controller" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["configmaps"], APIGroups:[""], Verbs:["watch"]} PolicyRule{Resources:["endpoints"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["endpoints"], APIGroups:[""], Verbs:["watch"]} PolicyRule{Resources:["nodes"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["nodes"], APIGroups:[""], Verbs:["watch"]} PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["pods"], APIGroups:[""], Verbs:["watch"]} PolicyRule{Resources:["secrets"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["secrets"], APIGroups:[""], Verbs:["watch"]} PolicyRule{Resources:["nodes"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["list"]} PolicyRule{Resources:["services"], APIGroups:[""], Verbs:["watch"]} PolicyRule{Resources:["ingresses"], APIGroups:["extensions"], Verbs:["get"]} PolicyRule{Resources:["ingresses"], APIGroups:["extensions"], Verbs:["list"]} PolicyRule{Resources:["ingresses"], APIGroups:["extensions"], Verbs:["watch"]} PolicyRule{Resources:["ingresses/status"], APIGroups:["extensions"], Verbs:["update"]} PolicyRule{Resources:["events"], APIGroups:[""], Verbs:["create"]} PolicyRule{Resources:["events"], APIGroups:[""], Verbs:["patch"]}] user=&{sameer@bitnami.com  [system:authenticated] map[]} ownerrules=[PolicyRule{Resources:["selfsubjectaccessreviews"], APIGroups:["authorization.k8s.io"], Verbs:["create"]} PolicyRule{NonResourceURLs:["/api" "/api/*" "/apis" "/apis/*" "/healthz" "/swaggerapi" "/swaggerapi/*" "/version"], Verbs:["get"]}] ruleResolutionErrors=[]
Error from server (Forbidden): error when creating "STDIN": roles.rbac.authorization.k8s.io "sealed-secrets-key-admin" is forbidden: attempt to grant extra privileges: [PolicyRule{Resources:["secrets"], ResourceNames:["sealed-secrets-key"], APIGroups:[""], Verbs:["get"]} PolicyRule{Resources:["secrets"], APIGroups:[""], Verbs:["create"]}] user=&{sameer@bitnami.com  [system:authenticated] map[]} ownerrules=[PolicyRule{Resources:["selfsubjectaccessreviews"], APIGroups:["authorization.k8s.io"], Verbs:["create"]} PolicyRule{NonResourceURLs:["/api" "/api/*" "/apis" "/apis/*" "/healthz" "/swaggerapi" "/swaggerapi/*" "/version"], Verbs:["get"]}] ruleResolutionErrors=[]
```

**Comment by sameersbn** [comment]
FWIW, this does not work

```
$ gcloud container clusters create sameer-cluster --no-enable-legacy-authorization
$ kubeapps up --dry-run | kubectl create -f -
```

But this works,

```
$ gcloud container clusters create sameer-cluster --no-enable-legacy-authorization
$ kubeapps up && kubeapps down
$ kubeapps up --dry-run | kubectl delete -f - # "kubeapps down" does not remove the clusterrolebindings
$ kubeapps up --dry-run | kubectl create -f -
```

**Comment by prydonius** [comment]
@sameersbn can you expand on why it doesn't work? Is it an RBAC issue? It worked fine for my on Docker for Mac's Kubernetes environment.

**Comment by sameersbn** [comment]
It's my understanding that the cluster roles did not get created for some reason. But I cannot be 100% sure. In your tests, did you use GKE?

**Comment by ngtuna** [comment]
@sameersbn @prydonius Yes the cluster role won't get created. The steps of `kubeapps up` are: 1) generating the inline clusterrolebinding object, 2) creating it, 3) deploying kubeapps, 4) revoking it immediately. 

So with `--dry-run` option, right now user has to create it manually. I will send another PR to take that step into the process.

**Pull request merged**